proxy/server.go

```diff
@@ -0,0 +1,139 @@
+package proxy
+
+import (
+	"errors"
+	"fmt"
+	"github.com/nuts-foundation/go-did/vc"
+	client "github.com/nuts-foundation/nuts-demo-ehr/client/auth"
+	"github.com/nuts-foundation/nuts-demo-ehr/domain/auth"
+	"github.com/nuts-foundation/nuts-node/vcr/credential"
+	"github.com/sirupsen/logrus"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+
+	"github.com/labstack/echo/v4"
+)
+
+type Server struct {
+	proxy *httputil.ReverseProxy
+	auth  auth.Service
+	path  string
+}
+
+func NewServer(authService auth.Service, targetURL url.URL, path string) *Server {
+	// Does not support query parameters in targetURL
+	proxyDirector := func(req *http.Request) {
+		req.URL.Scheme = targetURL.Scheme
+		req.URL.Host = targetURL.Host
+		req.URL.RawPath = "" // Not required?
+		req.URL.Path = targetURL.Path + req.URL.Path[len(path):]
+		logrus.Infof("Rewritten to: %s", req.URL.Path)
+		if _, ok := req.Header["User-Agent"]; !ok {
+			// explicitly disable User-Agent so it's not set to default value
+			req.Header.Set("User-Agent", "")
+		}
+	}
+	return &Server{
+		path:  path,
+		proxy: &httputil.ReverseProxy{Director: proxyDirector},
+		auth:  authService,
+	}
+}
+
+func findNutsAuthorizationCredential(token *client.TokenIntrospectionResponse) (*vc.VerifiableCredential, error) {
+	if token.Vcs != nil {
+		for _, verifiableCredential := range *token.Vcs {
+			types := credential.ExtractTypes(verifiableCredential)
+
+			for _, typeName := range types {
+				if typeName == credential.NutsAuthorizationCredentialType {
+					return &verifiableCredential, nil
+				}
+			}
+		}
+	}
+
+	return nil, errors.New("NutsAuthorizationCredential was not found in the access-token")
+}
+
+func (server *Server) parseAccessToken(c echo.Context) (*client.TokenIntrospectionResponse, error) {
+	bearerToken, err := server.auth.ParseBearerToken(c.Request())
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse the bearer token: %w", err)
+	}
+
+	token, err := server.auth.IntrospectAccessToken(c.Request().Context(), bearerToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to introspect token: %w", err)
+	}
+
+	if !token.Active {
+		return nil, fmt.Errorf("access-token is not active")
+	}
+
+	return token, nil
+}
+
+func (server *Server) verifyAccess(route *fhirRoute, token *client.TokenIntrospectionResponse) error {
```

Comment:
the `purposeOfUse`/`service` field is not used to select the correct verifier. For eOverdracht this will either be `eOverdracht-receiver` or `eOverdracht-receiver`.